### PR TITLE
User Login/Sign Up.

### PR DESCRIPTION
**Pull request by abhikanojia**

### CODE REVIEW

File: config/routes.rb [hunk 1]
```diff
@@ -1,3 +1,15 @@
...
+    post 'login' => 'sessions#create'
+    get 'logout' => 'sessions#destroy'
+  end
+
+  resources :users do
+    member do
+      get :confirm_email
+    end
+  end
+  resource :profile
```
Comment (sawangupta92) [inline config/routes.rb]: why profile?

File: db/migrate/20180105183110_create_users.rb [hunk 2]
```diff
@@ -0,0 +1,18 @@
+class CreateUsers < ActiveRecord::Migration[5.0]
+  def change
+    create_table :users do |t|
+      t.string :email, null: false
+      t.string :firstname
+      t.string :lastname
+      t.string :location
+      t.string :picture_path
```
Comment (sawangupta92) [inline db/migrate/20180105183110_create_users.rb]: How did you created these columns? I can't see paperclip in the gemfile

File: config/environments/development.rb [hunk 3]
```diff
@@ -51,4 +51,16 @@
   # Use an evented file watcher to asynchronously detect changes in source code,
   # routes, locales, etc. This feature depends on the listen gem.
   config.file_watcher = ActiveSupport::EventedFileUpdateChecker
+
+  config.action_mailer.delivery_method = :smtp
+
+  config.action_mailer.smtp_settings = {
+    :address => "smtp.gmail.com",
```
Comment (sawangupta92) [inline config/environments/development.rb]: use new hash syntax

File: app/controllers/profiles_controller.rb [hunk 4]
```diff
@@ -0,0 +1,13 @@
+class ProfilesController < ApplicationController
```
Comment (sawangupta92) [inline app/controllers/profiles_controller.rb]: Why do we need this controller?

File: app/controllers/profiles_controller.rb [hunk 5]
```diff
@@ -0,0 +1,13 @@
...
+  before_action :authorize_user
+  before_action :set_user
+
+  def show
+  end
+
+  private
+
+    def set_user
```
Comment (sawangupta92) [inline app/controllers/profiles_controller.rb]: We already have current_user, then why do we need to set this again here.

File: app/controllers/profiles_controller.rb [hunk 6]
```diff
@@ -0,0 +1,13 @@
+class ProfilesController < ApplicationController
+  before_action :authorize_user
```
Comment (sawangupta92) [inline app/controllers/profiles_controller.rb]: Why are you calling it here? if it has already been called in the application controller

File: app/controllers/sessions_controller.rb [hunk 7]
```diff
@@ -0,0 +1,41 @@
+class SessionsController < ApplicationController
+  skip_before_action :authorize_user
+  rescue_from ActiveRecord::RecordNotFound, with: :invalid_user
```
Comment (sawangupta92) [inline app/controllers/sessions_controller.rb]: Why are we raising any exception? It is a costly operation

File: app/controllers/sessions_controller.rb [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+      session[:user_id] = @user.id
+      redirect_to profile_path, success: 'Logged In Successfully.'
+    else
+      redirect_to login_path, error: 'Invalid Email/Password Combination.'
+    end
+  end
+
+  def destroy
+    if session[:user_id]
+      session[:user_id] = nil
```
Comment (sawangupta92) [inline app/controllers/sessions_controller.rb]: remove complete session

File: app/controllers/sessions_controller.rb [hunk 9]
```diff
@@ -0,0 +1,41 @@
+class SessionsController < ApplicationController
+  skip_before_action :authorize_user
+  rescue_from ActiveRecord::RecordNotFound, with: :invalid_user
+  before_action :set_user, :ensure_user_is_active, only: :create
```
Comment (sawangupta92) [inline app/controllers/sessions_controller.rb]: ensure if user is present in logout.

File: app/controllers/sessions_controller.rb [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+
+  def create
+    if @user && @user.try(:authenticate, params[:password])
+      session[:user_id] = @user.id
+      redirect_to profile_path, success: 'Logged In Successfully.'
```
Comment (sawangupta92) [inline app/controllers/sessions_controller.rb]: use I18n everywhere.

File: app/controllers/sessions_controller.rb [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+  end
+
+  private
+
+    def set_user
+      @user = User.find_by!(email: params[:email])
+    end
+
+    def ensure_user_is_active
+      if !@user.is_active? && !@user.activated_at?
```
Comment (sawangupta92) [inline app/controllers/sessions_controller.rb]: use unless

File: app/controllers/sessions_controller.rb [hunk 12]
```diff
@@ -0,0 +1,41 @@
...
+        redirect_to login_path, error: 'Please confirm your account before logging in.'
+      elsif !@user.is_active? && @user.activated_at?
+        redirect_to login_path, error: 'Your account has been deactivated by site administrator.'
+      end
+    end
+
+    def invalid_user
```
Comment (sawangupta92) [inline app/controllers/sessions_controller.rb]: should not be needed.

File: app/controllers/sessions_controller.rb (hunk 11 above)
Comment (sawangupta92) [inline app/controllers/sessions_controller.rb]: These are two different before_actions.

File: app/controllers/users_controller.rb [hunk 13]
```diff
@@ -0,0 +1,66 @@
+class UsersController < ApplicationController
+  before_action :set_user, only: :update
+  skip_before_action :authorize_user
+  rescue_from ActiveRecord::RecordNotFound, with: :invalid_user
+
+  # GET /users/new
```
Comment (sawangupta92) [inline app/controllers/users_controller.rb]: Do not use scaffolding.

File: app/controllers/users_controller.rb [hunk 14]
```diff
@@ -0,0 +1,66 @@
+class UsersController < ApplicationController
```
Comment (sawangupta92) [inline app/controllers/users_controller.rb]: same comments as above controller

File: app/controllers/sessions_controller.rb [hunk 15]
```diff
@@ -0,0 +1,41 @@
+class SessionsController < ApplicationController
+  skip_before_action :authorize_user
+  rescue_from ActiveRecord::RecordNotFound, with: :invalid_user
+  before_action :set_user, :ensure_user_is_active, only: :create
+
+  def create
+    if @user && @user.try(:authenticate, params[:password])
```
Comment (sawangupta92) [inline app/controllers/sessions_controller.rb]: No need to check if @user is present.

File: app/controllers/users_controller.rb [hunk 16]
```diff
@@ -0,0 +1,66 @@
...
+  end
+
+  # POST /users
+  # POST /users.json
+  def create
+    @user = User.new(user_params)
+
+    respond_to do |format|
+      if @user.save
+        UserMailer.verify_email(@user).deliver_now
```
Comment (sawangupta92) [inline app/controllers/users_controller.rb]: Send email from models.

File: app/controllers/users_controller.rb [hunk 17]
```diff
@@ -0,0 +1,66 @@
...
+        format.html { redirect_to login_path, success: 'You need to confirm your email to activate your account.' }
```
Comment (sawangupta92) [inline app/controllers/users_controller.rb]: I18N

Comment (sawangupta92) [inline app/controllers/users_controller.rb]: Are we using both formats?

File: app/controllers/users_controller.rb [hunk 18]
```diff
@@ -0,0 +1,66 @@
...
+
+  # PATCH/PUT /users/1
+  # PATCH/PUT /users/1.json
+  def update
+    respond_to do |format|
+      if @user.update(user_params)
+        format.html { redirect_to @user, notice: 'User was successfully updated.' }
+        format.json { render :show, status: :ok, location: @user }
+      else
+        format.html { render :edit }
```
Comment (sawangupta92) [inline app/controllers/users_controller.rb]: Do we show any error here also?

File: app/controllers/users_controller.rb [hunk 19]
```diff
@@ -0,0 +1,66 @@
...
+        format.json { render json: @user.errors, status: :unprocessable_entity }
+      end
+    end
+  end
+
+  # user verification through email
+  def confirm_email
+    user = User.find_by!(confirmation_token: params[:id])
```
Comment (sawangupta92) [inline app/controllers/users_controller.rb]: Do not use bang operation everywhere.

File: app/controllers/users_controller.rb [hunk 20]
```diff
@@ -0,0 +1,66 @@
...
+    end
+  end
+
+  private
+    # Use callbacks to share common setup or constraints between actions.
+    def set_user
+      @user = User.find(params[:id])
+    end
+
+    # Never trust parameters from the scary internet, only allow the white list through.
```
Comment (sawangupta92) [inline app/controllers/users_controller.rb]: Remove scaffolding comments

File: app/controllers/users_controller.rb [hunk 21]
```diff
@@ -0,0 +1,66 @@
...
+    def user_params
+      params.require(:user).permit(:firstname, :lastname, :email, :password, :password_confirmation)
```
Comment (sawangupta92) [inline app/controllers/users_controller.rb]: Is email updatable?

File: app/controllers/users_controller.rb [hunk 22]
```diff
@@ -0,0 +1,66 @@
...
+    end
+
+    def invalid_user
```
Comment (sawangupta92) [inline app/controllers/users_controller.rb]: No need.

File: app/helpers/application_helper.rb [hunk 23]
```diff
@@ -1,2 +1,31 @@
 module ApplicationHelper
+  def flash_class(level)
```
Comment (sawangupta92) [inline app/helpers/application_helper.rb]: From where did you get this module? or have you written it yourself?

File: app/helpers/application_helper.rb [hunk 24]
```diff
@@ -1,2 +1,31 @@
...
+    end
+  end
+
+  def current_user
+    if session[:user_id]
+      @current_user ||= User.find_by(id: session[:user_id])
+    end
+  end
+
+  def user_logged_in?
```
Comment (sawangupta92) [inline app/helpers/application_helper.rb]: what if I directly check for user_logged_in? w/o calling current_user method

File: app/mailers/user_mailer.rb [hunk 25]
```diff
@@ -0,0 +1,13 @@
...
+
+  # Subject can be set in your I18n file at config/locales/en.yml
+  # with the following lookup:
+  #
+  #   en.user_mailer.verify.subject
+  #
+  def verify_email(user)
+    @greeting = "Hi"
+    @user = user
+    mail to: @user.email
```
Comment (sawangupta92) [inline app/mailers/user_mailer.rb]: from?

File: app/models/user.rb [hunk 26]
```diff
@@ -0,0 +1,25 @@
+class User < ApplicationRecord
+  EMAIL_REGEX = /\A[\w+\-.]+@[a-z\d\-.]+\.[a-z]+\z/i
+
+  has_secure_password
+  validates :firstname, :lastname, :password, :email, presence: true
```
Comment (sawangupta92) [inline app/models/user.rb]: what if I want to update my firstname? do I have to enter my password?

File: app/models/user.rb [hunk 27]
```diff
@@ -0,0 +1,25 @@
...
+  validates :email, uniqueness: { case_sensitive: false }, format: { with: EMAIL_REGEX, multiline: true }, if: :email?
```
Comment (sawangupta92) [inline app/models/user.rb]: You can also use allow_nil here.

File: app/models/user.rb [hunk 28]
```diff
@@ -0,0 +1,25 @@
...
+
+
+  before_create :add_confirmation_token
```
Comment (sawangupta92) [inline app/models/user.rb]: rename

File: app/models/user.rb [hunk 29]
```diff
@@ -0,0 +1,25 @@
...
+
+  def activate
+    unless is_active
+      self.is_active = true
+      self.confirmation_token = nil
+      self.activated_at = Time.current.to_datetime
+      save!(validate: false)
```
Comment (sawangupta92) [inline app/models/user.rb]: why validate false?

File: app/models/user.rb [hunk 30]
```diff
@@ -0,0 +1,25 @@
...
+  EMAIL_REGEX = /\A[\w+\-.]+@[a-z\d\-.]+\.[a-z]+\z/i
+
+  has_secure_password
+  validates :firstname, :lastname, :password, :email, presence: true
+  validates :email, uniqueness: { case_sensitive: false }, format: { with: EMAIL_REGEX, multiline: true }, if: :email?
+
+
+  before_create :add_confirmation_token
+
+  def activate
```
Comment (sawangupta92) [inline app/models/user.rb]: rename to activate!

File: app/views/sessions/create.html.erb [hunk 31]
```diff
@@ -0,0 +1,2 @@
+<h1>Sessions#create</h1>
```
Comment (sawangupta92) [inline app/views/sessions/create.html.erb]: Remove these scaffolding files.

File: app/helpers/application_helper.rb (hunk 23 above)
Comment (abhikanojia) [reply app/helpers/application_helper.rb]: This module was generated by rails itself . I have added the methods.

File: app/controllers/users_controller.rb (hunk 16 above)
Comment (abhikanojia) [reply app/controllers/users_controller.rb]: Fixed in remember_me branch.

File: app/controllers/application_controller.rb [hunk 32]
```diff
@@ -1,3 +1,14 @@
...
+  add_flash_types :success, :warning, :danger, :info, :error
   protect_from_forgery with: :exception
+  before_action :authorize_user
+
+  protected
+
+    def authorize_user
+      @current_user ||= User.find_by(id: session[:user_id])
+      unless @current_user
+        redirect_to login_path, notice: 'Please login to continue.'
```
Comment (sawangupta92) [inline app/controllers/application_controller.rb]: use I18n everywhere in controllers and models.

File: app/controllers/sessions_controller.rb [hunk 33]
```diff
@@ -0,0 +1,41 @@
...
+  def create
+    if @user.try(:authenticate, params[:password])
+      session[:user_id] = @user.id
+      redirect_to user_path(@user), success: t('.login_success')
+    else
+      redirect_to login_path, error: t('.login_error')
+    end
+  end
+
+  def destroy
```
Comment (sawangupta92) [inline app/controllers/sessions_controller.rb]: authorize_user here

File: app/controllers/sessions_controller.rb [hunk 34]
```diff
@@ -0,0 +1,41 @@
...
+
+  private
+
+    def set_user
+      @user = User.find_by(email: params[:email])
+      redirect_to(login_path, error: t('.user_not_found')) unless @user
+    end
+
+    def ensure_user_is_active
+      unless @user.is_active? && @user.activated_at?
```
Comment (sawangupta92) [inline app/controllers/sessions_controller.rb]: move this logic in model

File: app/controllers/sessions_controller.rb [hunk 35]
```diff
@@ -0,0 +1,41 @@
...
+        redirect_to login_path, error: t('.confirm_account_msg')
+      end
+    end
+
+    def ensure_user_not_disabled_by_admin
+      if !@user.is_active? && @user.activated_at?
```
Comment (sawangupta92) [inline app/controllers/sessions_controller.rb]: move this logic in model

File: app/controllers/users_controller.rb [hunk 36]
```diff
@@ -0,0 +1,52 @@
...
+  before_action :ensure_user_is_accessing_its_account, :set_user, only: :show
+
+  def new
+    @user = User.new
+  end
+
+  def create
+    @user = User.new(user_params)
+
+    respond_to do |format|
```
Comment (sawangupta92) [inline app/controllers/users_controller.rb]: why respond_to if we are only working with html?

File: app/controllers/users_controller.rb [hunk 37]
```diff
@@ -0,0 +1,52 @@
...
+      params.require(:user).permit(:firstname, :lastname, :email, :password, :password_confirmation)
+    end
+
+    def set_user_by_confirmation_token
+      @user = User.find_by(confirmation_token: params[:id])
+      redirect_to(login_path, error: t('.user_not_found')) unless @user.present?
+    end
+
+    def ensure_user_is_accessing_its_account
+      redirect_to(user_path(@current_user), warning: t('.not_allowed')) unless params[:id].to_i.eql? @current_user.id
```
Comment (sawangupta92) [inline app/controllers/users_controller.rb]: this line is becoming too long, break it using block

File: app/controllers/users_controller.rb [hunk 38]
```diff
@@ -0,0 +1,52 @@
+class UsersController < ApplicationController
+  skip_before_action :authorize_user, except: :show
+  before_action :set_user_by_confirmation_token, only: :confirm_email
+  before_action :ensure_user_is_accessing_its_account, :set_user, only: :show
```
Comment (sawangupta92) [inline app/controllers/users_controller.rb]: I think we can view anyone's profile

File: app/mailers/user_mailer.rb (hunk 25 above)
Comment (sawangupta92) [inline app/mailers/user_mailer.rb]: subject?

File: app/mailers/user_mailer.rb [hunk 39]
```diff
@@ -0,0 +1,13 @@
+class UserMailer < ApplicationMailer
+
+  # Subject can be set in your I18n file at config/locales/en.yml
+  # with the following lookup:
+  #
+  #   en.user_mailer.verify.subject
+  #
+  def verify_email(user)
+    @greeting = "Hi"
```
Comment (sawangupta92) [inline app/mailers/user_mailer.rb]: why this?

File: app/models/user.rb [hunk 40]
```diff
@@ -0,0 +1,25 @@
+class User < ApplicationRecord
+  EMAIL_REGEX = /\A[\w+\-.]+@[a-z\d\-.]+\.[a-z]+\z/i
```
Comment (sawangupta92) [inline app/models/user.rb]: Rails provides email validation

File: app/controllers/users_controller.rb [hunk 41]
```diff
@@ -0,0 +1,52 @@
...
+        format.html { render :new }
+      end
+    end
+  end
+
+  def show
+  end
+
+  def confirm_email
+    if user.activate
```
Comment (sawangupta92) [inline app/controllers/users_controller.rb]: method name is activate! then how are we calling activate here?

File: app/views/layouts/application.html.erb [hunk 42]
```diff
@@ -8,7 +8,69 @@
     <%= javascript_include_tag 'application', 'data-turbolinks-track': 'reload' %>
   </head>
 
+  <nav class="navbar navbar-default">
+    <div class="container-fluid">
+      <div class="navbar-header">
```
Comment (sawangupta92) [inline app/views/layouts/application.html.erb]: move header in a separate file

File: app/views/layouts/application.html.erb [hunk 43]
```diff
@@ -8,7 +8,69 @@
     <%= javascript_include_tag 'application', 'data-turbolinks-track': 'reload' %>
   </head>
 
+  <nav class="navbar navbar-default">
```
Comment (sawangupta92) [inline app/views/layouts/application.html.erb]: Why it is not in body??

File: app/views/sessions/destroy.html.erb [hunk 44]
```diff
@@ -0,0 +1,2 @@
+<h1>Sessions#destroy</h1>
```
Comment (sawangupta92) [inline app/views/sessions/destroy.html.erb]: Please, delete these files

File: app/views/sessions/new.html.erb [hunk 45]
```diff
@@ -0,0 +1,39 @@
+<%= render partial: 'shared/flash' %>
```
Comment (sawangupta92) [inline app/views/sessions/new.html.erb]: render this in application.html file

File: app/views/shared/_error.html.erb [hunk 46]
```diff
@@ -0,0 +1,9 @@
+<% if object.errors.any? %>
+  <div id="error_explanation">
+    <ul>
+    <% object.errors.full_messages.each do |message| %>
```
Comment (sawangupta92) [inline app/views/shared/_error.html.erb]: Indentation

File: app/views/user_mailer/verify_email.text.erb [hunk 47]
```diff
@@ -0,0 +1,3 @@
+User#verify
+
+<%= @greeting %>, find me in app/views/user_mailer/verify.text.erb
```
Comment (sawangupta92) [inline app/views/user_mailer/verify_email.text.erb]: use pro[er text

File: app/views/users/index.html.erb [hunk 48]
```diff
@@ -0,0 +1,25 @@
+<p id="notice"><%= notice %></p>
```
Comment (sawangupta92) [inline app/views/users/index.html.erb]: Why???

File: config/routes.rb [hunk 49]
```diff
@@ -1,3 +1,14 @@
 Rails.application.routes.draw do
+  controller :sessions do
+    get 'login' => 'sessions#new'
+    post 'login' => 'sessions#create'
+    get 'logout' => 'sessions#destroy'
```
Comment (sawangupta92) [inline config/routes.rb]: delete action

File: config/routes.rb [hunk 50]
```diff
@@ -1,3 +1,14 @@
 Rails.application.routes.draw do
+  controller :sessions do
```
Comment (sawangupta92) [inline config/routes.rb]: use only: [] also

File: app/controllers/sessions_controller.rb [hunk 51]
```diff
@@ -0,0 +1,41 @@
+class SessionsController < ApplicationController
+  skip_before_action :authorize_user, except: :destroy
+  before_action :set_user, :ensure_user_is_active, :ensure_user_not_disabled_by_admin, only: :create
+
+  def create
+    if @user.try(:authenticate, params[:password])
```
Comment (sawangupta92) [inline app/controllers/sessions_controller.rb]: why try? how will it be possible that this does not have @user if we already have set_user before action.

File: app/controllers/sessions_controller.rb [hunk 52]
```diff
@@ -0,0 +1,41 @@
...
+      session[:user_id] = @user.id
+      redirect_to user_path(@user), success: t('.login_success')
+    else
+      redirect_to login_path, error: t('.login_error')
+    end
+  end
+
+  def destroy
+    if session[:user_id]
```
Comment (sawangupta92) [inline app/controllers/sessions_controller.rb]: if we have found the current_user already that means session does have a user_id, then why are we checking it again here?

File: app/mailers/user_mailer.rb [hunk 53]
```diff
@@ -0,0 +1,12 @@
+class UserMailer < ApplicationMailer
+
+  # Subject can be set in your I18n file at config/locales/en.yml
```
Comment (sawangupta92) [inline app/mailers/user_mailer.rb]: remove these comments

File: app/mailers/user_mailer.rb [hunk 54]
```diff
@@ -0,0 +1,12 @@
...
+  # with the following lookup:
+  #
+  #   en.user_mailer.verify.subject
+  #
+  def verify_email(user)
+    @user = user
+    mail to: @user.email
```
Comment (sawangupta92) [inline app/mailers/user_mailer.rb]: subject?

File: app/models/user.rb [hunk 55]
```diff
@@ -0,0 +1,38 @@
...
+
+  has_secure_password
+  validates :firstname, :lastname, :email, presence: true
+  validates :email, presence: true, uniqueness: { case_sensitive: false }, format: { with: EMAIL_REGEX, multiline: true }, allow_nil: true
```
Comment (sawangupta92) [inline app/models/user.rb]: presence: true twice on email

File: app/models/user.rb [hunk 56]
```diff
@@ -0,0 +1,38 @@
...
+  def account_is_active?
+    self.is_active? && self.activated_at?
+  end
+
+  def disabled_by_admin?
+    !(self.is_active?) && self.activated_at?
+  end
+
+  def send_verification_email
+    UserMailer.verify_email(self).deliver_now
```
Comment (sawangupta92) [inline app/models/user.rb]: send only id

File: app/views/sessions/new.html.erb [hunk 57]
```diff
@@ -0,0 +1,38 @@
...
+          </div>
+        </div>
+        <div class="form-group">
+          <div class="col-lg-10 col-lg-offset-2">
+            <%= button_tag 'Login', type: 'submit', class: 'btn btn-primary' %>
+            <%= button_tag 'Reset', type: 'reset', class: 'btn btn-default' %>
+          </div>
+        </div>
+      </fieldset>
+    </form>
```
Comment (sawangupta92) [inline app/views/sessions/new.html.erb]: ?? end use do end block on form_tag

File: config/locales/en.yml [hunk 58]
```diff
@@ -20,4 +20,25 @@
 # available at http://guides.rubyonrails.org/i18n.html.
 
 en:
-  hello: "Hello world"
+  sessions:
+    create:
+      login_success: 'Logged In Successfully.'
```
Comment (sawangupta92) [inline config/locales/en.yml]: rename key to success and error

File: app/controllers/sessions_controller.rb [hunk 59]
```diff
@@ -0,0 +1,37 @@
...
+  def destroy
+    reset_session
+    redirect_to login_path, success: t('.logout_success')
+  end
+
+  private
+
+    def set_user
+      @user = User.find_by(email: params[:email])
+      redirect_to(login_path, error: t('.user_not_found')) unless @user
```
Comment (sawangupta92) [inline app/controllers/sessions_controller.rb]: let's discuss

File: app/helpers/users_helper.rb [hunk 60]
```diff
@@ -0,0 +1,5 @@
+module UsersHelper
+  def user_full_name
```
Comment (sawangupta92) [inline app/helpers/users_helper.rb]: rename to full_name

File: app/mailers/user_mailer.rb [hunk 61]
```diff
@@ -0,0 +1,14 @@
+class UserMailer < ApplicationMailer
+  ACCOUNT_VERIFICATION = 'Account Verification'
+
+  def verify_email(user_id)
+    @user = set_user(user_id)
+    mail to: @user.email, subject: ACCOUNT_VERIFICATION
```
Comment (sawangupta92) [inline app/mailers/user_mailer.rb]: No need to use constant here, if you really want then you may use i18n.

File: app/mailers/user_mailer.rb [hunk 62]
```diff
@@ -0,0 +1,14 @@
...
+  end
+
+  private
+
+    def set_user(user_id)
```
Comment (sawangupta92) [inline app/mailers/user_mailer.rb]: why a separate method?

File: app/controllers/users_controller.rb [hunk 63]
```diff
@@ -0,0 +1,44 @@
...
+  skip_before_action :authorize_user, except: :show
+  before_action :set_user_by_confirmation_token, only: :confirm_email
+  before_action :set_user, only: :show
+
+  def new
+    @user = User.new
+  end
+
+  def create
+    @user = User.new(user_params)
```
Comment (sawangupta92) [inline app/controllers/users_controller.rb]: User.user.new

File: app/models/user.rb [hunk 64]
```diff
@@ -0,0 +1,39 @@
...
+      self.activated_at = Time.current.to_datetime
+      save!
+    end
+  end
+
+  def account_is_active?
+    self.is_active? && self.activated_at?
+  end
+
+  def disabled_by_admin?
```
Comment (sawangupta92) [inline app/models/user.rb]: rename to is_disabled?

File: app/views/users/new.html.erb [hunk 65]
```diff
@@ -0,0 +1,49 @@
...
+      <div class="form-group">
+        <div class="col-lg-10 col-lg-offset-2">
+          <%= button_tag 'Sign Up', type: 'submit', class: 'btn btn-primary' %>
+          <%= button_tag 'Reset', type: 'reset', class: 'btn btn-default' %>
+        </div>
+      </div>
+    <% end %>
+  </div>
+  <div class="panel-footer">
+    <a href="<%= login_path %>">Already a User? Login</a>
```
Comment (sawangupta92) [inline app/views/users/new.html.erb]: link_to

**Pull request merged**